MAINT: Add a fast path to var for complex input

rossbar opened the pull request:
`var` currently has a conditional that results in `conjugate` being called for the variance calculation of complex inputs.

This leg of the computation is slower than it could be if `conjugate` were avoided altogether. This PR avoids this computational leg for complex inputs via a type check.

Closes #15684.

Here are the results of a benchmark on my system:

On 6894bbc6 (i.e. pre-change):
```
             =========== =============
                 arr size                
              ----------- -------------
                   10        35.3±1μs  
                  100      35.3±0.05μs 
                  1000       41.1±2μs  
                 10000       72.9±5μs  
                 100000     725±300μs  
                1000000     8.21±0.9ms 
                10000000     94.0±3ms  
               100000000     898±10ms  
              =========== =============
```

This PR (with fast path):
```
             =========== =============
                 arr size                
              ----------- -------------
                   10       77.7±0.3μs 
                  100      77.6±0.08μs 
                  1000       85.7±4μs  
                 10000       111±7μs   
                 100000      339±2μs   
                1000000      6.03±1ms  
                10000000     75.4±6ms  
               100000000     756±30ms  
              =========== =============
```

I'm not sure adding the additional complexity is necessarily worth the speed-up, especially since the change results in a slow-down for the non-computation-constrained case (i.e. smaller arrays) - please let me know what you think.

rossbar committed:
ENH: Adds a fast path to var for complex input

var currently has a conditional that results in conjugate being
called for the variance calculation of complex inputs.

This leg of the computation is slow. This PR avoids this
computational leg for complex inputs via a type check.

Closes #15684

rossbar commented:
Related: #13177

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -165,6 +165,15 @@ def _mean(a, axis=None, dtype=None, out=None, keepdims=False):
+        return um.add(
+            _var(arr.real, axis, dtype, out, ddof, keepdims),
+            _var(arr.imag, axis, dtype, out, ddof, keepdims)
+        )
```
Reusing `out` here doesn't seem safe

rossbar replied:
Good point - will look into it and think about adding new test(s) for this extra path.

In your opinion, is the change even worth it? I'm not sure how I feel about adding extra complexity for ~20% speedup (which is very system-dependent).

eric-wieser replied:
Can you get the same speedup by adding the fast path right where `conjugate` is called?

rossbar replied:
I had originally been attempting a solution along those lines - If there were a `conjmultiply` ufunc for instance that would do the trick but I don't know of any such function.

eric-wieser replied:
How about
```
um.multiply(x.real, x.real, out=x.real)
um.multiply(x.imag, x.imag, out=x.imag)
x = np.add(x.real, x.imag)  # could use out=x.real, but may be slower
```

rossbar replied:
I will try along those lines - thanks for the suggestion

rossbar replied:
`um.add(um.multiply(x.real, x.real), um.multiply(x.imag, x.imag))`

This results in the speedup, but unfortunately still implicitly depends on the dtype of the input being complex - it fails on the test suite for object arrays with complex numbers in them

eric-wieser replied:
I think special casing complex dtypes is ok, as long as you do it in the same place that the other types are specialized.

rossbar replied:
Thanks for the feedback - the current approach (c049074) keeps all of the special cases based on dtypes in the same area of the code where all the other type-determination occurs.

rossbar committed:
Added benchmark for _var with complex input.

rossbar commented:
I've updated the PR to incorporate @eric-wieser 's feedback about keeping the conditions based on dtypes all contained in the same part of the code. The most recent version adds fast-paths for `complex128` and `complex64` input.

I also added two new tests: one to test `_var` for single precision numbers, and a second for testing `_var` for >2 dimensional arrays as the fast paths use a `view` of the orignal array that increases the number of dimensions.

Finally, since this PR is about performance, I added a benchmark to `bench_core`. I'm not sure whether modifying the benchmarks is encouraged, but it should at least help evaluate the PR. If it shouldn't be included, I'm happy to revert (or you can - the benchmark is added in c049074).

Thanks to @seberg  for the `view` idea and a lot of other great input re: ufunc behavior.

Here are the current results of the benchmark on my system:

**Original implementation**:
```
              =========== =============
                 arr size                
              ----------- -------------
                   10       40.3±0.4μs 
                  100       40.2±0.9μs 
                  1000     46.8±0.03μs 
                 10000       94.0±7μs  
                 100000     585±300μs  
                1000000    8.15±0.06ms 
                10000000     91.4±2ms  
               100000000     914±8ms   
              =========== =============
```
**Manually computing var from real & imag components `a.real.var() + a.imag.var()`**:
```
              =========== =============
                 arr size                
              ----------- -------------
                   10        75.0±2μs  
                  100       74.8±0.3μs 
                  1000       83.4±3μs  
                 10000       116±4μs   
                 100000      411±10μs  
                1000000    6.10±0.05ms 
                10000000    73.6±0.7ms 
               100000000     745±7ms   
              =========== =============
```
**`_var` fast path (this PR)**:
```
              =========== ============
                 arr size               
              ----------- ------------
                   10      43.4±0.2μs 
                  100      44.0±0.1μs 
                  1000      50.0±1μs  
                 10000      72.5±3μs  
                 100000     286±7μs   
                1000000    5.88±0.2ms 
                10000000    66.7±3ms  
               100000000    674±4ms   
              =========== ============
```

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -191,6 +191,17 @@ def _var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=False):
+        xv = x.view(dtype=(nt.float64, (2,)))
+        um.multiply(xv, xv, out=xv)
+        x = um.add(xv[..., 0], xv[..., 1], out=x.real).real
+    elif arr.dtype.type is nt.complex64:
```
Can you add `clongdouble` too? Perhaps add a dictionary to map the types.

seberg replied:
Agreed, one thing to take care of (and that should have a test) is that non-native byte order does not use the view, or only uses it with the correct byte order. (The identity check does not catch byte order differences. An equality check, even if a bit strange maybe, does). Not sure of the top of my head how best to write it nicest.
After my PR goes in, we could even think about adding a `.real_dtype()` method ;).

rossbar replied:
Thanks @eric-wieser & @seberg for the suggestions! I've made the following changes/updates:

 - Added a fast-path for `complex256`.
 - Re-formulated the complex type checking to use a dict (eliminates some LOC's, doesn't appear to hurt the performance)
 - Added an additional check for the byteorder - any array that has anything other than non-native byte order will be funnelled to the original "slow" path so that it doesn't hit the `view`s.
 - Updated the tests:
   * Modified a test to check `complex{64,128,256}` input types
   * Added a test to verify `var()` gives the same result for complex arrays with both native and non-native byte order.

Let me know what you think of this iteration, and thanks again for the feedback!

seberg commented on numpy/core/_methods.py:
```diff
@@ -189,8 +189,25 @@ def _var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=False):
+    _complex_to_float = {
+        nt.complex64 : nt.float32,
+        nt.complex128 : nt.float64,
+        nt.clongdouble: nt.longdouble
```
```suggestion
        nt.clongdouble: nt.longdouble,
```

seberg commented on numpy/core/_methods.py:
```diff
@@ -189,8 +189,25 @@ def _var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=False):
+    # Fast-paths for built-in complex types with native byte order
+    elif (
+        x.dtype.type in _complex_to_float.keys() and arr.dtype.isnative
+    ):
```
```suggestion
    elif x.dtype.type in _complex_to_float and arr.dtype.isnative:
```
Seems it is not even above 80, but even if it was, I think in this case we should probably not care...

seberg commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -5507,6 +5507,40 @@ def test_var_values(self):
+        cmat = self.cmat.copy().astype('complex128')
+        for bo_modifier in ('<', '>'):
+            carr = np.array(cmat, dtype=bo_modifier + 'c16')
+            assert_almost_equal(carr.var(), cmat.var())
```
```suggestion
        cmat_swapped = cmat.astype(cmat.dtype.newbyteorder())
        assert_almost_equal(cmat.var(), cmat_swapped.var())
```

seberg commented on benchmarks/benchmarks/bench_core.py:
```diff
@@ -180,3 +180,17 @@ def time_unpackbits_axis1_little(self):
+        self.arr.var()
+
+    def time_var_real_imag_independently(self, n):
+        self.arr.real.var() + self.arr.imag.var()
```
```suggestion

```

seberg commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -5507,6 +5507,40 @@ def test_var_values(self):
+    @pytest.mark.parametrize(('complex_dtype', 'ndec'), (
+        ('complex64', 6),
+        ('complex128', 7),
+        ('clongdouble', 7)
```
```suggestion
        ('clongdouble', 7),
```

seberg commented on numpy/core/_methods.py:
```diff
@@ -189,8 +189,25 @@ def _var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=False):
+    elif (
+        x.dtype.type in _complex_to_float.keys() and arr.dtype.isnative
+    ):
+        # Computation for fast-path
```
```suggestion
```

seberg reviewed (approved):
Generally looks good to me, small fixes. Will probably look over it briefly and merge soon if nobody has another opinion.

seberg commented:
Can you add the benchmark as the first commit, and then squash the rest into a single commit?

rossbar commented:
Thanks for all the feedback @seberg - it should be down to two commits now: one with the updated code/tests, and a separate one for the accompanying benchmark.

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -189,8 +189,25 @@ def _var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=False):
+    _complex_to_float = {
+        nt.complex64 : nt.float32,
+        nt.complex128 : nt.float64,
+        nt.clongdouble: nt.longdouble
```
If you make the keys `np.dtype` instances, and crucially reverse the order, then you can avoid the `isnative` check below.

eric-wieser replied:
Reversing is necessary so that when double and longdouble are equivalent, the latter, double, wins.

rossbar replied:
I don't understand the comment on ordering; particularly the second comment. Also, I thought that Python dictionaries didn't make any guarantees about ordering. Would this need to be an OrderedDict?

eric-wieser replied:
Dictionary literals are evaluated sequentially. double and long double are considered equal on Windows, you want double to take precedence.

eric-wieser replied:
Apologies for the lack of explanation, will expand when not on Mobile.

eric-wieser replied:
Perhaps best to keep things explicit:
```python
_complex_to_float = {
    np.dtype(np.csingle): np.dtype(np.single),
    np.dtype(np.cdouble): np.dtype(np.double),
}
# on windows these are equivalent, and we want double to take precedence
if np.dtype(np.longdouble) != np.dtype(np.double):
    _complex_to_float.update({
    	np.dtype(np.clongdouble): np.dtype(np.longdouble),
    })
# add the reverse-endian types
_complex_to_float.update({
	k.newbyteorder(): v.newbyteorder()
	for k, v in _complex_to_float.items()
})
```

---

The remark I was making above is that:
```python
d = {
    1: "a",
    2: "b",
    1: "c"
}
```
is treated as equivalent to
```python
d = {}
d[1] = "a"
d[2] = "b"
d[1] = "c"
```
eg when repeated, the last repetition wins.

rossbar replied:
I see now, thanks for the explanation - one thing however: I understood from @seberg 's comment that non-native types are to be excluded, so we wouldn't want to add the reverse byteorders to the fast-path dictionary

eric-wieser replied:
I think @seberg meant that viewing a big-endian `complex double` as a little-endian `double` was unsafe. As long as you match the endianness before and after the view, everything will be fine.

rossbar replied:
Thanks for the clarification :+1:

rossbar replied:
One issue with the concrete dtype instances is that attempting to import `dtype` from `numpy.core` or the full namespace (`import numpy as np`) results in a circular import in `numpy/core/_methods.py`. Is there a way to get concrete datatype instances using only `numpy.core.numerictypes`?

eric-wieser replied:
`np.core.numerictypes.dtype`?

rossbar replied:
Thanks, I don't know why I had such a hard time chasing that down :expressionless:

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -189,8 +189,25 @@ def _var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=False):
     x = asanyarray(arr - arrmean)
+
+    # Complex types to -> (2,)float view for fast-path
+    _complex_to_float = {
```
May as well make this global, no need to construct it every time.

rossbar replied:
I had considered that as well - the overhead from building the dictionary seemed negligible so I had left it in, but I can just as well move it outside the function :+1:

eric-wieser replied:
Let's move it out, if nothing else if makes `var` shorter and easier to read.

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -189,8 +189,22 @@ def _var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=False):
         x = um.multiply(x, x, out=x)
+    # Fast-paths for built-in complex types with native byte order
+    elif x.dtype.type in _complex_to_float and arr.dtype.isnative:
+        xv = x.view(dtype=(_complex_to_float[x.dtype.type], (2,)))
```
With my dict suggestion, this becomes
```suggestion
    elif x.dtype in _complex_to_float:
        xv = x.view(dtype=(_complex_to_float[x.dtype], (2,)))
```

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -5507,6 +5507,39 @@ def test_var_values(self):
+
+    def test_var_dimensions(self):
+        # _var paths for complex number introduce additions on views that
+        # increase dimensions. Ensure this generalizes to higher dims
```
Note that this will make `var` fail on 32-dimensional arrays, since you'll try to insert another dimension which isn't allowed. I don't know if we care.

rossbar replied:
We could add another condition of `arr.ndim < 32` to the fast-branch of the conditional if you think it's worth it. Is `32` hard-coded as the max dim in NumPy or is that a configurable value?

eric-wieser replied:
Let's see what @seberg thinks. It's probably ok not to bother.

seberg replied:
Lets not bother, if you got 31 dimensions there are already a lot of other operations that break.

rossbar commented:
Thanks for the additional feedback @eric-wieser, the latest commit has incorporated your suggestions:
 * Moved the definition of the complex->float mapping outside of `_var` and use concrete dtype instances
 * Explicitly handle the double/longdouble case for increased code clarity
 * Added non-native byteorders to the mapping and modified the corresponding type check in `_var`

Everything should still be formulated so that the code/tests are in one commit (c898ff3) and the benchmark is in another (ffe1f46).

Let me know if I've missed anything.

seberg commented:
Thanks Ross and Eric, I am going to put this in. I doubt the byteswapped paths are really worth it speed wise, but it doesn't add bad complexity either and keeps things symmetric.

eric-wieser commented:
> I doubt the byteswapped paths are really worth it speed wise

I'm starting to think they're unreachable, since the subtraction output is probably always native...

seberg commented:
Hmmpf, true should have noticed... I guess ross will fix it up probably.
